### ISSUE TITLE
Large numbers of namespaces * languages use higher CPU than necessary - O(n^2)

### ISSUE THREAD

**Issue by huntharo** [i18next/i18next#1767]
## 🐛 Bug Report

`O(n^2)` CPU usage on the number of `namespaces * languages` passed to `BackendConnector.load()`.

On a particular application I work on there are ~20 languages and ~400 namespaces with 8,000 total files.  The application is using `next-i18next` with the `i18next-fs-backend`.  In this case the loading of the files takes ~4 seconds total on a Linux docker image running on Mac / EC2 / Lambda / etc.  We are aiming to reduce this as much as possible and noticed that the loader code is using arrays and `indexOf` calls on those arrays on the total number of `languages * namespaces`.

The fix in #1768 reduces the loading time for 10,000 items to 2% of what it was before and changes the O(n^2) relationship to O(n).

## To Reproduce

- 1,000 items takes `0.083s` on a particular machine
- 10,000 items takes `6.97s` on that same machine
  - 84x longer duration for 10x more input

### 1,000 Items Baseline

<img width="929" alt="image" src="https://user-images.githubusercontent.com/5617868/167280180-db815418-af17-4b4b-ac83-f39329e796de.png">

### 10,000 Items

<img width="923" alt="image" src="https://user-images.githubusercontent.com/5617868/167280184-d230c3ea-0e70-4535-aa11-032ce19da8e0.png">

```js
import BackendConnector from '../../src/BackendConnector.js';
import BackendMock from './backendMock.js';
import Interpolator from '../../src/Interpolator.js';
import ResourceStore from '../../src/ResourceStore.js';
import {expect} from 'chai';

describe('BackendConnector performance test', () => {
  let connector;

  before(() => {
    connector = new BackendConnector(
      new BackendMock(),
      new ResourceStore(),
      {
        interpolator: new Interpolator(),
      },
      {
        backend: { loadPath: 'http://localhost:9876/locales/{{lng}}/{{ns}}.json' },
      },
    );
  });

  describe('#load', () => {
    it('should load 10,000 items in under the 2 second timeout', (done) => {
      const namespaces = [];
      for (let i = 0; i < 10000; i++) {
        namespaces.push(`namespace${i}`);
      }
      connector.load(['en'], namespaces, function (err) {
        expect(err).to.be.not.ok;
        expect(connector.store.getResourceBundle('en', 'namespace1')).to.eql({
          status: 'nok',
          retries: 0,
        });
        done();
      });
    });
  });
});
```

## Expected behavior

- The computation time should grow linearly relative to the `languages * namespaces` count
- Processing `10,000` items should take less time than it does currently

### 1,000 Items Baseline

<img width="989" alt="image" src="https://user-images.githubusercontent.com/5617868/167280402-5720f811-cbaa-4251-97af-a32d329096e1.png">

### 10,000 Items

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/5617868/167280413-386ec77b-874a-4566-9bd2-7786d2696a5a.png">

## Your Environment

- *runtime version*: node v14
- *i18next version*: repo clone on May 7, 2022
- *os*: Mac / Linux
- *any other relevant information*

**Comment by adrai** [comment]
wow, a lot of namespaces....
[PR](https://github.com/i18next/i18next/pull/1768) looks good to me...
regarding #1769 I've proposed a [less invasive approach](https://github.com/i18next/i18next/issues/1769#issuecomment-1120357392)...

**Comment by jamuhl** [comment]
Just out of curiosity...I'm wondering what system needs to load that amount of translations file. Do you have a namespace per component? Do you have to load every language and every application section on init?

I know larger apps that might get to a few hundred files with lazy loading as needed...but this numbers...I have difficulties to see what that system could be using such numbers.

Further how whose numbers of file get translated...looks like a lot of pain...

**Comment by huntharo** [comment]
@jamuhl I am not positive on what drove the namespace count but yes I think it may be one per component. I will see if I can get some of the folks involved in that pattern to share the reasoning.

We load the namespaces on init partly because of the unbounded concurrency and CPU usage from loading large numbers. We noticed that for the first few minutes after startup the app would stall. We moved init to startup in an attempt to limit down delays on requests for a ready instance.

With the CPU usage reduction we may be able to remove the init on startup.

Hmm... now that I think about it I think the O(n^2) problem would have happened 100 times if 100 requests for the same namespaces happened before they finished loading. I need to check that but that could explain the improvement we saw in moving loading to startup.
